DIPs/1NNN-JH-TG-NT.md

```diff
@@ -0,0 +1,483 @@
+
+`out` can be applied to the whole unpack parameter, but not to individual fields. `auto ref` and `lazy` cannot
+be applied on an unpacking parameter list.
+
+Note: `out` unpacking parameters are not supported by the current implementation, but may be added at a future date.
+
+
+### Grammar changes
+The following grammar additions will be sufficient.
+(Note that many grammar rules are repeated due to (existing) distinct restrictions on what `StorageClasses` are allowed; this can make the proposed grammar changes look more profound than they are.)
+
+#### Unpacking declarations
+See <https://dlang.org/spec/declaration.html#VarDeclarations>.
+```diff
+  VarDeclarations:
+      ...
+      AutoDeclaration
++     StorageClasses[opt] TupleDeclarators ;
+
++ TupleDeclarators:
++     TupleDeclarator
++     TupleDeclarator , TupleDeclarators
+
++ TupleDeclarator:
++     ( TupleDeclarators2 ) = Initializer
+
++ TupleDeclarators2:
++     TupleDeclarator2 ,
++     TupleDeclarator2 , TupleDeclarator2
++     TupleDeclarator2 , TupleDeclarators2
+
++ TupleDeclarator2:
++     StorageClasses[opt] Identifier
++     StorageClasses[opt] BasicType TypeSuffixes[opt] Identifier
++     StorageClasses[opt] ( TupleDeclarators2 )
+```
+
+#### Unpacking `foreach` elements
+(Note that `ForeachTupleDeclarators2` is the same as `TupleDeclarators2` except that the latter allows all `StorageClasses`, while the former allows only `ForeachTypeAttributes`.)
+
+See <https://dlang.org/spec/statement.html#ForeachType>.
+```diff
+  ForeachType:
+      ...
+      ForeachTypeAttributes[opt] alias Identifier
++     ForeachTypeAttributes[opt] ( ForeachTupleDeclarators2 )
+
++ ForeachTupleDeclarators2:
++     ForeachTupleDeclarator2 ,
++     ForeachTupleDeclarator2 , ForeachTupleDeclarator2
++     ForeachTupleDeclarator2 , ForeachTupleDeclarators2
+
++ ForeachTupleDeclarator2:
++     ForeachTypeAttributes[opt] Identifier
++     ForeachTypeAttributes[opt] BasicType TypeSuffixes[opt] Identifier
++     ForeachTypeAttributes[opt] ( ForeachTupleDeclarators2 )
+```
+
+#### Unpacking function literal parameters
+(Note that `ParameterTupleDeclarator2` is the same as `TupleDeclarator2` except that the latter allows all `StorageClasses`, while the former allows only `ParameterAttributes`.)
+
+See <https://dlang.org/spec/function.html#Parameter>.
+```diff
+Parameter:
+      ...
+      ParameterDeclaration = AssignExpression ...
++     ParameterAttributes[opt] ParameterTupleDeclarator
+
++ ParameterTupleDeclarator:
++     ( ParameterTupleDeclarators2 )
++     ( ParameterTupleDeclarators2 ) = Initializer
+
++ ParameterTupleDeclarators2:
++     ParameterTupleDeclarator2 ,
++     ParameterTupleDeclarator2 , ParameterTupleDeclarator2
++     ParameterTupleDeclarator2 , ParameterTupleDeclarators2
+
++ ParameterTupleDeclarator2:
++     ParameterAttributes[opt] Identifier
++     ParameterAttributes[opt] BasicType TypeSuffixes[opt] Identifier
++     ParameterAttributes[opt] ( ParameterTupleDeclarators2 )
```

Comment:
Here, `ParameterTupleDeclarator(s)2` is a quite uninformative name. I suggest `ParameterTupleInnerDeclarator(s)`. (This doesn’t change what the grammar defines.)
```suggestion
+     ParameterAttributes[opt] ParameterTupleDeclarator

+ ParameterTupleDeclarator:
+     ( ParameterTupleInnerDeclarators )
+     ( ParameterTupleInnerDeclarators ) = Initializer

+ ParameterTupleInnerDeclarators:
+     ParameterTupleInnerDeclarator ,
+     ParameterTupleInnerDeclarator , ParameterTupleInnerDeclarator
+     ParameterTupleInnerDeclarator , ParameterTupleInnerDeclarators

+ ParameterTupleInnerDeclarator:
+     ParameterAttributes[opt] Identifier
+     ParameterAttributes[opt] BasicType TypeSuffixes[opt] Identifier
+     ParameterAttributes[opt] ( ParameterTupleInnerDeclarators )
```